File: DynamicProgramming.go
package main

import "fmt"

var counterPlainRecursive int
var counterDynamicProgramming int

func fibonacciPlainRecursive(n int) int {
	counterPlainRecursive++
	if n < 2 {
		return n
	}
	return fibonacciPlainRecursive(n-1) + fibonacciPlainRecursive(n-2)
}

func fibonacciDynamicProgramming() func(int) int {
	var cache = make(map[int]int)
	var fib func(n int) int
	fib = func(n int) int {
		counterDynamicProgramming++
		if val, ok := cache[n]; ok {
			return val
		} else {
			if n < 2 {
				return n
			} else {
				val := fib(n-1) + fib(n-2)
				cache[n] = val
				return val
			}
		}
	}
	return fib
}

func main() {
	fibonacciPlainRecursive(20)
	fasterFibonacci := fibonacciDynamicProgramming()
	fasterFibonacci(20)
	fmt.Printf("we did %v calculations for Plain Recursive\n", counterPlainRecursive)
	fmt.Printf("we did %v calculations for Dynamic Programming\n", counterDynamicProgramming)
}

File: DynamicProgramming.kt
var counterPlainRecursive = 0
var counterDynamicProgramming = 0

fun fibonacciPlainRecursive(n: Int): Int {
    counterPlainRecursive++
    return if (n < 2) {
        n
    } else {
        fibonacciPlainRecursive(n - 1) + fibonacciPlainRecursive(n - 2)
    }
}

fun fibonacciDynamicProgramming(): (Int) -> Int {
    val cache = mutableMapOf<Int, Int>()
    fun fib(n: Int): Int {
        counterDynamicProgramming++
        if (n in cache) {
            return cache[n]!!
        } else {
            val result = if (n < 2) {
                n
            } else {
                fib(n - 1) + fib(n - 2)
            }
            cache[n] = result
            return cache[n]!!
        }
    }
    return ::fib
}

fun main(){
fibonacciPlainRecursive(20)
val fasterFibonacci = fibonacciDynamicProgramming()
fasterFibonacci(20)
println("we did $counterPlainRecursive calculations for Plain Recursive")
println("we did $counterDynamicProgramming calculations for Dynamic Programming")
}